pkg/policy/correlation/correlation.go

```diff
@@ -0,0 +1,173 @@
+// provided flow.
+func CorrelatePolicy(endpointGetter getters.EndpointGetter, f *flowpb.Flow) {
+	if f.GetEventType().GetType() != int32(monitorAPI.MessageTypePolicyVerdict) ||
+		f.GetVerdict() != flowpb.Verdict_FORWARDED {
+		// we are only interested in policy verdict notifications for forwarded flows
+		return
+	}
+
+	// extract fields relevant for looking up the policy
+	direction, endpointIP, remoteIdentity, proto, dport := extractFlowKey(f)
+
+	// obtain reference to endpoint on which the policy verdict was taken
+	epInfo, ok := endpointGetter.GetEndpointInfo(endpointIP)
+	if !ok {
+		logger.WithField(logfields.IPAddr, endpointIP).Debug("failed to lookup endpoint")
+		return
+	}
+
+	derivedFrom, rev, ok := lookupPolicyForKey(epInfo, policy.Key{
+		Identity:         uint32(remoteIdentity),
+		DestPort:         dport,
+		Nexthdr:          uint8(proto),
+		TrafficDirection: uint8(direction),
+	})
+	if !ok {
+		logger.WithFields(logrus.Fields{
+			logfields.Identity:         remoteIdentity,
+			logfields.Port:             dport,
+			logfields.Protocol:         proto,
+			logfields.TrafficDirection: direction,
+		}).Debug("unable to find policy for policy verdict notification")
+		return
+	}
+
+	allowedBy := toProto(derivedFrom, rev)
+	switch direction {
+	case trafficdirection.Egress:
+		f.EgressAllowedBy = allowedBy
+	case trafficdirection.Ingress:
+		f.IngressAllowedBy = allowedBy
+	}
+}
+
+func extractFlowKey(f *flowpb.Flow) (
+	direction trafficdirection.TrafficDirection,
+	endpointIP netip.Addr,
+	remoteIdentity identity.NumericIdentity,
+	proto u8proto.U8proto,
+	dport uint16) {
+
+	switch f.GetTrafficDirection() {
+	case flowpb.TrafficDirection_EGRESS:
+		direction = trafficdirection.Egress
+		endpointIP, _ = netip.ParseAddr(f.GetIP().GetSource())
+		remoteIdentity = identity.NumericIdentity(f.GetDestination().GetIdentity())
+	case flowpb.TrafficDirection_INGRESS:
+		direction = trafficdirection.Ingress
+		endpointIP, _ = netip.ParseAddr(f.GetIP().GetDestination())
+		remoteIdentity = identity.NumericIdentity(f.GetSource().GetIdentity())
+	default:
+		direction = trafficdirection.Invalid
+		endpointIP = netip.IPv4Unspecified()
+		remoteIdentity = identity.IdentityUnknown
+	}
+
+	if tcp := f.GetL4().GetTCP(); tcp != nil {
+		proto = u8proto.TCP
+		dport = uint16(tcp.GetDestinationPort())
+	} else if udp := f.GetL4().GetUDP(); udp != nil {
+		proto = u8proto.UDP
+		dport = uint16(udp.GetDestinationPort())
+	} else {
+		proto = u8proto.ANY
+		dport = 0
+	}
+
+	return
+}
+
+func lookupPolicyForKey(ep v1.EndpointInfo, key policy.Key) (derivedFrom labels.LabelArrayList, rev uint64, ok bool) {
+	// Check for L4 policy rules
+	derivedFrom, rev, ok = ep.GetRealizedPolicyRuleLabelsForKey(key)
+
+	// TODO: We should inspect f.GetPolicyMatchType() to reduce the number of lookups here
+	if !ok {
+		// Check for L3 policy rules
+		derivedFrom, rev, ok = ep.GetRealizedPolicyRuleLabelsForKey(policy.Key{
```

Comment:
Make sure to check that `DestPort` and `NextHdr` aren't already `0`, as it is a possibility in the `extractFlowKey` method.